File: src/differential_equations_separable_equations_hard.py
"""
differential_equations - separable_equations (hard)
Generated: 2026-02-22T05:19:41.525814
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)

    if problem_type == 1:
        # IVP: y^2 dy = k*x^(m-1) dx
        k = nonzero(-5, 5)
        m = choice([2, 3, 4, 5])
        x0 = choice([0, 1, 2])
        y0 = nonzero(1, 5)

        C_val = Rational(y0**3, 3) - k * Rational(x0**m, m)

        question = f"Solve the IVP: $\\frac{{dy}}{{dx}} = \\frac{{{latex(k * x**(m-1))}}}{{y^2}}$, $y({x0}) = {y0}$"

        ans = cbrt(3 * k * x**m / m + 3 * C_val)

        sol_steps = steps(
            f"Separate variables: $y^2\\,dy = {latex(k * x**(m-1))}\\,dx$",
            f"Integrate: $\\frac{{y^3}}{{3}} = {latex(k * x**m / m)} + C$",
            f"Apply $y({x0}) = {y0}$: $C = {latex(C_val)}$",
            f"$y^3 = {latex(3 * k * x**m / m + 3 * C_val)}$",
            f"$y = {latex(ans)}$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1650, 1750),
            topic="differential_equations/separable_equations",
            solution=sol_steps
        )

    elif problem_type == 2:
        # Logistic equation: dy/dx = y(a - b*y)
        a = randint(1, 5)
        b = randint(1, 4)

        question = f"Solve: $\\frac{{dy}}{{dx}} = y({a} - {b}y)$"

        C1 = symbols('C_1')
        ans = a * C1 * exp(a * x) / (1 + b * C1 * exp(a * x))

        sol_steps = steps(
            f"Separate: $\\frac{{dy}}{{y({a} - {b}y)}} = dx$",
            f"Partial fractions: $\\frac{{1}}{{y({a} - {b}y)}} = \\frac{{1}}{{{a}y}} + \\frac{{{b}}}{{{a}({a} - {b}y)}}$",
            f"Integrate: $\\frac{{1}}{{{a}}}\\ln|y| - \\frac{{1}}{{{a}}}\\ln|{a} - {b}y| = x + C$",
            f"Simplify: $\\ln\\left|\\frac{{y}}{{{a} - {b}y}}\\right| = {a}x + {a}C$",
            f"Exponentiate: $\\frac{{y}}{{{a} - {b}y}} = C_1 e^{{{a}x}}$",
            f"Solve for $y$: $y = \\frac{{{a}C_1 e^{{{a}x}}}}{{1 + {b}C_1 e^{{{a}x}}}}$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1700, 1800),
            topic="differential_equations/separable_equations",
            solution=sol_steps
        )

    elif problem_type == 3:
        # dy/dx = f(x) * sec^2(y)  =>  cos^2(y) dy = f(x) dx  =>  tan(y) = F(x) + C
        a = nonzero(-5, 5)
        m = choice([1, 2, 3])
        f_x = a * x**m
        F_x = integrate(f_x, x)

        question = f"Solve: $\\frac{{dy}}{{dx}} = {latex(f_x)} \\sec^2(y)$"

        C1 = symbols('C_1')
        ans_implicit = Eq(tan(y), F_x + C1)

        sol_steps = steps(
            f"Separate: $\\frac{{dy}}{{\\sec^2(y)}} = {latex(f_x)}\\,dx$, i.e. $\\cos^2(y)\\,dy = {latex(f_x)}\\,dx$",
            f"Since $\\frac{{d}}{{dy}}[\\tan y] = \\sec^2 y$, divide both sides of original by $\\sec^2 y$:",
            f"$\\int dy/\\sec^2 y$ leads to $\\tan(y) = \\int {latex(f_x)}\\,dx$",
            f"$\\tan(y) = {latex(F_x)} + C_1$"
        )

        return problem(
            question=question,
            answer=ans_implicit,
            difficulty=(1600, 1700),
            topic="differential_equations/separable_equations",
            solution=sol_steps,
            answer_type="equation"
        )

    elif problem_type == 4:
        # Limited-growth IVP: dy/dt = k*y*(M - y)
        k_choices = [Rational(1, 10), Rational(1, 5), Rational(1, 4), Rational(1, 8), Rational(3, 20)]
        k = choice(k_choices)
        M = choice([50, 100, 200, 150])
        y0 = choice([5, 10, 20, 25])
        t0 = 0

        K_val = Rational(y0, M - y0)

        ans = M * K_val * exp(M * k * t) / (1 + K_val * exp(M * k * t))

        question = f"Solve the IVP: $\\frac{{dy}}{{dt}} = {latex(k)}y({M} - y)$, $y({t0}) = {y0}$"

        sol_steps = steps(
            f"Separate: $\\frac{{dy}}{{y({M} - y)}} = {latex(k)}\\,dt$",
            f"Partial fractions: $\\frac{{1}}{{y({M} - y)}} = \\frac{{1}}{{{M}y}} + \\frac{{1}}{{{M}({M} - y)}}$",
            f"Integrate and exponentiate: $\\frac{{y}}{{{M} - y}} = Ke^{{{latex(M * k)}t}}$",
            f"Apply $y({t0}) = {y0}$: $K = \\frac{{{y0}}}{{{M - y0}}} = {latex(K_val)}$",
            f"$y = {latex(ans)}$"
        )

        return problem(
            question=question,
            answer=ans,
            difficulty=(1750, 1850),
            topic="differential_equations/separable_equations",
            solution=sol_steps
        )

    else:
        # dy/dx = e^(a*x + b*y)
        a = nonzero(-4, 4)
        b = nonzero(-4, 4)

        question = f"Solve: $\\frac{{dy}}{{dx}} = e^{{{latex(a * x + b * y)}}}$"

        C1 = symbols('C_1')
        # e^(-b*y) dy = e^(a*x) dx
        # -1/b * e^(-b*y) = 1/a * e^(a*x) + C
        ans_implicit = Eq(-exp(-b * y) / b, exp(a * x) / a + C1)

        sol_steps = steps(
            f"Factor: $\\frac{{dy}}{{dx}} = e^{{{a}x}} \\cdot e^{{{b}y}}$",
            f"Separate: $e^{{-{b}y}}\\,dy = e^{{{a}x}}\\,dx$",
            f"Integrate: $-\\frac{{e^{{-{b}y}}}}{{{b}}} = \\frac{{e^{{{a}x}}}}{{{a}}} + C_1$"
        )

        return problem(
            question=question,
            answer=ans_implicit,
            difficulty=(1650, 1760),
            topic="differential_equations/separable_equations",
            solution=sol_steps,
            answer_type="equation"
        )


emit(generate())

File: src/differential_equations_separable_equations_hard.jl
# differential_equations - separable_equations (hard)
# Generated: 2026-03-08T20:50:04.835390

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y t begin
    set_topic!("differential_equations/separable_equations")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Type 1: dy/dx = f(x)g(y) with polynomial factors
        a = nonzero(-8, 8)
        b = nonzero(-8, 8)
        k = nonzero(-6, 6)
        
        # dy/dx = k*f(x)*g(y)
        f_x = a*x + b
        g_y = y^2 + nonzero(-15, 15)
        
        rhs = k * f_x * g_y
        
        # Integrate: ∫dy/(y²+c) = ∫k(ax+b)dx
        # (1/√c)arctan(y/√c) = k(ax²/2 + bx) + C
        
        problem(
            question="Solve the separable differential equation: \$\\frac{dy}{dx} = $(tex(rhs))\$",
            answer="separable",
            difficulty=(1800, 2100),
            grading_mode="text",
            solution=steps(
                sol("Given", diff(y, x) ~ rhs),
                "Separate variables: \$\\frac{dy}{$(tex(g_y))} = $(k) \\cdot $(tex(f_x)) \\, dx\$",
                "Integrate both sides",
                "Solution involves arctan or logarithmic terms depending on the discriminant"
            ),
            time=240,
            calculator="none"
        )
        
    elseif problem_type == 2
        # Type 2: dy/dx = y/x type (homogeneous)
        a = nonzero(-7, 7)
        b = nonzero(-7, 7)
        c = nonzero(-7, 7)
        
        # dy/dx = (ay + bx)/(cx)
        numerator = a*y + b*x
        denominator = c*x
        
        problem(
            question="Solve the separable differential equation: \$\\frac{dy}{dx} = \\frac{$(tex(numerator))}{$(tex(denominator))}\$",
            answer="separable",
            difficulty=(1900, 2200),
            grading_mode="text",
            solution=steps(
                sol("Given", diff(y, x) ~ numerator/denominator),
                "Rewrite as: \$\\frac{dy}{dx} = \\frac{$(a)y}{$(c)x} + \\frac{$(b)}{$(c)}\$",
                "Separate: \$\\frac{dy}{$(a)y + $(b*c)x/$(c)} = \\frac{dx}{$(c)x}\$",
                "Integrate both sides to obtain implicit solution"
            ),
            time=270,
            calculator="none"
        )
        
    elseif problem_type == 3
        # Type 3: dy/dx = f(x)/g(y) clean separation
        m = nonzero(-6, 6)
        n = nonzero(-6, 6)
        p = nonzero(-5, 5)
        q = nonzero(-5, 5)
        
        # dy/dx = (mx + n)/(py + q)
        numerator = m*x + n
        denominator = p*y + q
        
        problem(
            question="Find the general solution to \$\\frac{dy}{dx} = \\frac{$(tex(numerator))}{$(tex(denominator))}\$",
            answer="separable",
            difficulty=(1850, 2050),
            grading_mode="text",
            solution=steps(
                sol("Given", diff(y, x) ~ numerator/denominator),
                "Separate variables: \$($(tex(denominator))) \\, dy = ($(tex(numerator))) \\, dx\$",
                "Integrate: \$\\int ($(p)y + $(q)) \\, dy = \\int ($(m)x + $(n)) \\, dx\$",
                "General solution: \$\\frac{$(p)y^2}{2} + $(q)y = \\frac{$(m)x^2}{2} + $(n)x + C\$"
            ),
            time=210,
            calculator="none"
        )
        
    elseif problem_type == 4
        # Type 4: Exponential growth/decay with initial condition
        k = nonzero(-8, 8)
        y0 = nonzero(-20, 20)
        t0 = rand([0, 1])
        
        # dy/dt = ky, y(t0) = y0
        
        problem(
            question="Solve the initial value problem: \$\\frac{dy}{dt} = $(k)y\$, \$y($(t0)) = $(y0)\$",
            answer="exponential",
            difficulty=(1800, 2000),
            grading_mode="text",
            solution=steps(
                sol("Given IVP", diff(y, t) ~ k*y),
                "Separate and integrate: \$\\int \\frac{dy}{y} = \\int $(k) \\, dt\$",
                "General solution: \$\\ln|y| = $(k)t + C\$, so \$y = Ae^{$(k)t}\$",
                "Apply initial condition \$y($(t0)) = $(y0)\$: \$A = $(y0)e^{$(-k*t0)}\$",
                "Particular solution: \$y = $(y0)e^{$(k)(t - $(t0))}\$"
            ),
            time=240,
            calculator="none"
        )
        
    else
        # Type 5: More complex separable with trig or rational functions
        a = nonzero(-6, 6)
        b = nonzero(-6, 6)
        c = nonzero(-4, 4)
        
        # dy/dx = (ax² + b)/(y)
        numerator = a*x^2 + b
        denominator = y
        
        y0 = nonzero(-15, 15)
        x0 = rand([0, 1, -1])
        
        problem(
            question="Solve the initial value problem: \$\\frac{dy}{dx} = \\frac{$(tex(numerator))}{y}\$, \$y($(x0)) = $(y0)\$",
            answer="separable",
            difficulty=(2000, 2300),
            grading_mode="text",
            solution=steps(
                sol("Given IVP", diff(y, x) ~ numerator/denominator),
                "Separate variables: \$y \\, dy = ($(tex(numerator))) \\, dx\$",
                "Integrate both sides: \$\\int y \\, dy = \\int ($(a)x^2 + $(b)) \\, dx\$",
                "General solution: \$\\frac{y^2}{2} = \\frac{$(a)x^3}{3} + $(b)x + C\$",
                "Apply \$y($(x0)) = $(y0)\$: find \$C = $(y0^2/2 - a*x0^3/3 - b*x0)\$",
                "Particular solution: \$y^2 = \\frac{2 \\cdot $(a)x^3}{3} + $(2*b)x + $(y0^2 - 2*a*x0^3/3 - 2*b*x0)\$"
            ),
            time=300,
            calculator="scientific"
        )
    end
end